File: leetcode-04/solutions/cpp/0714.cpp
class Solution {
 public:
  int maxProfit(vector<int>& prices, int fee) {
    int sell = 0;
    int hold = INT_MIN;

    for (int price : prices) {
      int prev = sell;
      sell = max(sell, hold + price);
      hold = max(hold, prev - price - fee);
    }

    return sell;
  }
};

File: leetcode-04/solutions/python3/0714.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        sell = 0
        hold = float('-inf')

        for price in prices:
            prev = sell
            sell = max(sell, hold + price)
            hold = max(hold, prev - price - fee)

        return sell
